pkg/pop/poptoken.go

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package pop
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"math/big"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+const popTokenType = "pop"
+
+type PopAuthenticationScheme struct {
+	// host is the u claim we will add on the pop token
+	Host   string
+	PoPKey PoPKey
+}
+
+func (as *PopAuthenticationScheme) TokenRequestParams() map[string]string {
+	return map[string]string{
+		"token_type": popTokenType,
+		"req_cnf":    as.PoPKey.ReqCnf(),
+	}
+}
+
+func (as *PopAuthenticationScheme) KeyID() string {
+	return as.PoPKey.KeyID()
+}
+
+func (as *PopAuthenticationScheme) FormatAccessToken(accessToken string) (string, error) {
+	ts := time.Now().Unix()
+	nonce := uuid.New().String()
+	nonce = strings.ReplaceAll(nonce, "-", "")
+	header := fmt.Sprintf(`{"typ":"%s","alg":"%s","kid":"%s"}`, popTokenType, as.PoPKey.Alg(), as.PoPKey.KeyID())
+	headerB64 := base64.RawURLEncoding.EncodeToString([]byte(header))
+	payload := fmt.Sprintf(`{"at":"%s","ts":%d,"u":"%s","cnf":{"jwk":%s},"nonce":"%s"}`, accessToken, ts, as.Host, as.PoPKey.JWK(), nonce)
+	payloadB64 := base64.RawURLEncoding.EncodeToString([]byte(payload))
+	h256 := sha256.Sum256([]byte(headerB64 + "." + payloadB64))
+	signature, err := as.PoPKey.Sign(h256[:])
+	if err != nil {
+		return "", err
+	}
+	signatureB64 := base64.RawURLEncoding.EncodeToString(signature)
+
+	return headerB64 + "." + payloadB64 + "." + signatureB64, nil
```

Comment:
Can we create a struct for this ?

Reply:
instead of manipulating string directly via `fmt.Sprintf()` in the function we can for sure get a representation of what is a pop token via a struct, something like :
```
type PoPAccessToken struct {
   Header header
   Payload payload
   Signature signature
}

type header struct {
  typ string
  alg string
  kid string
}

func (h *header) ToString() string {
  return fmt.Sprintf({"typ":"%s","alg":"%s","kid":"%s"}, h.typ, h.alg, h.kid)
}

func (h *header) ToBase64() string {
  base64.RawURLEncoding.EncodeToString([]byte(h.ToString()))
}
```

Reply:
Added structs to handle this